### PR TITLE
prioritized readable and writable iterators

### PR DESCRIPTION
**Pull request by LPardue**
This change modifies the internal implementaion of
readable and writable stream collections so that they
better respect the ordering expected from RFC 9218.

### CODE REVIEW

File: quiche/src/stream.rs [hunk 1]
```diff
@@ -943,14 +946,20 @@ impl Iterator for StreamIter {
 
     #[inline]
     fn next(&mut self) -> Option<Self::Item> {
-        self.streams.pop()
+        let ret = self.streams.get(self.index).cloned();
```
Comment (ghedo) [inline quiche/src/stream.rs:949]: Somewhat simpler would be:
```rust
let v = self.streams.get(self.index)?;
self.index += 1;
Some(*v)
```
Not 100% sure it compiles though.

File: quiche/src/stream.rs [hunk 2]
```diff
@@ -421,11 +432,42 @@ impl StreamMap {
...
+
+            self.writable_prioritized.insert(Arc::clone(priority_key));
         } else {
             self.writable.remove(&stream_id);
+
+            self.writable_prioritized_remove(priority_key);
+        }
+    }
+
+    fn writable_prioritized_remove(
```
Comment (ghedo) [inline quiche/src/stream.rs:450]:
```suggestion
    fn remove_writable(
```

File: quiche/src/stream.rs [hunk 3]
```diff
@@ -421,11 +432,42 @@ impl StreamMap {
...
+    ) {
+        let mut c = {
+            let ptr = Arc::as_ptr(priority_key);
+            unsafe { self.writable_prioritized.cursor_mut_from_ptr(ptr) }
+        };
+
+        c.remove();
+    }
+
+    pub fn writable_prioritized_update(
```
Comment (ghedo) [inline quiche/src/stream.rs:461]:
```suggestion
    pub fn update_priority(
```
(we can use this method in the future for readable/flushable too)

File: quiche/src/stream.rs [hunk 4]
```diff
@@ -741,17 +844,99 @@ pub fn is_bidi(stream_id: u64) -> bool {
...
+#[derive(Debug)]
+pub struct StreamPriorityKey {
+    pub urgency: u8,
+    pub incremental: bool,
+    pub id: u64,
+
+    pub writable: RBTreeAtomicLink,
+}
+
+impl Clone for StreamPriorityKey {
```
Comment (ghedo) [inline quiche/src/stream.rs:857]: Can't we just derive this?

File: quiche/src/lib.rs [hunk 5]
```diff
@@ -4670,10 +4680,26 @@ impl Connection {
             return Ok(());
         }
 
+        let is_writable = stream.is_writable();
+
         stream.urgency = urgency;
         stream.incremental = incremental;
 
-        // TODO: reprioritization
+        let old_priority_key = stream.priority_key.clone();
```
Comment (ghedo) [inline quiche/src/lib.rs:4688]: I wonder if something like the following would work:
```rust
let new_priority_key = Arc::new(...);
let old_priority_key = std::mem::replace(&mut stream.priority_key, new_priority_key);

self.streams.writable_prioritized_update(
  &old_priority_key,
  &stream.priority_key,
  is_writable,
);
```
Looks a little cleaner (if it works).

**Review by ghedo — CHANGES_REQUESTED** [review]
Added a few (minor) comments.

File: quiche/src/stream.rs [hunk 6]
```diff
@@ -421,11 +432,42 @@ impl StreamMap {
...
+
+    pub fn writable_prioritized_update(
+        &mut self, old: &Arc<StreamPriorityKey>, new: &Arc<StreamPriorityKey>,
+        writable: bool,
+    ) {
+        if old.writable.is_linked() {
+            self.writable_prioritized_remove(old);
+        }
+
+        if writable {
```
Comment (ghedo) [inline quiche/src/stream.rs:469]: If the old key was linked then we can assume the new one should be inserted too? This way we wouldn't need the additional `writable` argument.

File: quiche/src/lib.rs (hunk 5 above)
Comment (LPardue) [reply quiche/src/lib.rs]: yeah this works

File: quiche/src/stream.rs (hunk 4 above)
Comment (LPardue) [reply quiche/src/stream.rs]: yep

File: quiche/src/stream.rs [hunk 7]
```diff
@@ -678,6 +775,12 @@ impl Stream {
             local,
             urgency: DEFAULT_URGENCY,
             incremental: true,
+            priority_key: Arc::new(StreamPriorityKey {
```
[the remark this answers is not in the document]
Comment (LPardue) [reply quiche/src/stream.rs]: done

File: quiche/src/stream.rs [hunk 8]
```diff
@@ -421,11 +432,42 @@ impl StreamMap {
     /// to when an existing stream becomes writable (or stops being writable).
     ///
     /// If the stream was already in the list, this does nothing.
-    pub fn mark_writable(&mut self, stream_id: u64, writable: bool) {
+    pub fn mark_writable(
```
[the remark this answers is not in the document]
Comment (LPardue) [reply quiche/src/stream.rs:435]: Seemed easier to split them in a separate PR, see https://github.com/cloudflare/quiche/pull/1553

File: quiche/src/stream.rs (hunk 6 above)
Comment (ghedo) [reply quiche/src/stream.rs:473]: @LPardue it seems this comment still applies. Also, probably need to add the same logic for `readable` too.

File: quiche/src/lib.rs [hunk 9]
```diff
@@ -4435,7 +4439,7 @@ impl Connection {
         }
 
         if !readable {
-            self.streams.remove_readable(stream_id);
+            self.streams.remove_readable(&priority_key);
```
Comment (ghedo) [inline quiche/src/lib.rs:4442]: I suppose we need to do the shuffling for `readable` too? like:
```
        if incremental && readable {
            // Shuffle the incremental stream to the back of the the queue.
            self.streams.remove_readable(&priority_key);
            self.streams.insert_readable(&priority_key);
        }
```

**Review by ghedo — APPROVED** [review]

**Pull request merged**